library/pod-security-policy/read-only-root-filesystem/template.yaml

```diff
@@ -37,67 +37,104 @@ spec:
                 type: string
   targets:
     - target: admission.k8s.gatekeeper.sh
-      rego: |
-        package k8spspreadonlyrootfilesystem
+      code:
+      - engine: K8sNativeValidation
+        source:
+          variables:
+          - name: containers
+            expression: 'has(variables.anyObject.spec.containers) ? variables.anyObject.spec.containers : []'
+          - name: initContainers
+            expression: 'has(variables.anyObject.spec.initContainers) ? variables.anyObject.spec.initContainers : []'
+          - name: ephemeralContainers
+            expression: 'has(variables.anyObject.spec.ephemeralContainers) ? variables.anyObject.spec.ephemeralContainers : []'
+          - name: exemptImagePrefixes
+            expression: |
+              !has(variables.params.exemptImages) ? [] :
+                variables.params.exemptImages.filter(image, image.endsWith("*")).map(image, string(image).replace("*", ""))
+          - name: exemptImageExplicit
+            expression: |
+              !has(variables.params.exemptImages) ? [] : 
+                variables.params.exemptImages.filter(image, !image.endsWith("*"))
+          - name: exemptImages
+            expression: |
+              (variables.containers + variables.initContainers + variables.ephemeralContainers).filter(container,
+                container.image in variables.exemptImageExplicit ||
+                variables.exemptImagePrefixes.exists(exemption, string(container.image).startsWith(exemption)))
+          - name: badContainers
+            expression: |
+              (variables.containers + variables.initContainers + variables.ephemeralContainers).filter(container,
+                !(container.image in variables.exemptImages) && 
+                (!has(container.securityContext) ||
+                !has(container.securityContext.readOnlyRootFilesystem) ||
+                container.securityContext.readOnlyRootFilesystem != true)
+              ).map(container, container.name)
+          validations:
+          - expression: '(has(request.operation) && request.operation == "UPDATE") || size(variables.badContainers) == 0'
+            messageExpression: '"only read-only root filesystem container is allowed: " + variabales.badContainers.join(", ")'
```

Comment:
`variables` is misspelled.

Reply:
why didnt CI/tests fail?

Reply:
It seems that CEL was returning with the message for not recognizing the misspelled variable, but `gator` somehow treated it as violation (which it was expecting for disallow case). And since respective `suite,yaml` didn't specify expectation of specific message in case of violation, the error didn't get caught in CI.

Reply:
I fixed the variable and added verification of violation message in `suite.yaml`

Reply:
If there is an error executing `messageExpression`, VAP returns a default message instead.